### ISSUE TITLE
Reproduce Flex (Duo)'s training.

### ISSUE THREAD

**Issue by sagawatatsuya** [batistagroup/DirectMultiStep#17]
Hello,

I'm currently working on retraining the Flex (Duo) model from scratch and attempting to reproduce the results reported in the paper. I have modified use-examples/train-model.py as shown below.

However, the performance I obtain is significantly lower than the values reported in the paper. Could you kindly share the full training configuration or any additional details required to reproduce the reported results?

Training code used:
```python
from pathlib import Path

from directmultistep import helpers
from directmultistep.model import ModelFactory
from directmultistep.training import ModelTrainer, TrainingConfig

__mode__ = "local"
assert __mode__ in ["local", "cluster"]

if __mode__ == "local":
    # replace with .parent if you place train-model.py in root folder
    # .parents[1] if you keep it in use-examples
    base_path = Path(__file__).resolve().parents[1]
    n_workers = 8
    n_devices = 1
    accelerator = "cuda"
    batch_size = 32 * 4
elif __mode__ == "shared":
    base_path = Path(__file__).resolve().parent  # change it to your path
    n_workers = 64
    n_devices = 1
    accelerator = "auto"
    batch_size = 32 * 4
data_path = base_path / "data"

factory = ModelFactory.from_preset("flex_20M", compile_model=False)
# or any other preset name from src/directmultistep/dms/model/default_configs
# or create your own config, see src/directmultistep/model/factory.py for examples
config = TrainingConfig(
    data_path=data_path,
    run_name="reproduce_flex_duo_result",
    train_fname="/data1/sagawa/DirectMultiStep/data/processed/unique_dataset_nperms=3_nsms=all_noboth_train=0.95.pkl",
    val_fname="/data1/sagawa/DirectMultiStep/data/processed/unique_dataset_nperms=3_nsms=all_noboth_val=0.05.pkl",
    metadata_fname="/data1/sagawa/DirectMultiStep/data/configs/dms_dictionary.yaml",
    batch_size=batch_size,
    learning_rate=2e-4,
    max_epochs=40,
    warmup_steps=3000,
    decay_steps=80_000,
    decay_factor=0.1,
    pad_idx=factory.config.decoder.pad_idx,
    mask_idx=factory.config.decoder.mask_idx,
    save_top_k=-1,  # -1 will save all
    checkpoint_every_n_epochs=2,  # every 2 epochs
    summary_depth=2,
    accelerator=accelerator,
    num_workers=n_workers,
    n_devices=n_devices,
)

# Save configs to logbook
logbook_path = data_path / "configs" / "logbook" / config.run_name
logbook_path.mkdir(parents=True, exist_ok=True)

config.save(logbook_path / "training_config.yaml")
factory.config.save(logbook_path / "model_config.yaml")


train_dataset, val_dataset = helpers.prepare_datasets(
    train_data_path=data_path / "processed" / config.train_fname,
    val_data_path=data_path / "processed" / config.val_fname,
    metadata_path=data_path / "configs" / config.metadata_fname,
    load_sm=True,
    mode="training",
)

model = factory.create_model()
trainer = ModelTrainer(config)

if __name__ == "__main__":
    # this has to be in main block to avoid issues with multiprocessing
    trainer.train(model, train_dataset, val_dataset)
```
Inference code used:
```python
from pathlib import Path

from directmultistep.generation.eval import EvalConfig, ModelEvaluator
from directmultistep.model import ModelFactory
from directmultistep.training import TrainingConfig

__mode__ = "local"
assert __mode__ in ["local", "cluster"]

if __mode__ == "local":
    base_path = Path(__file__).resolve().parent.parent
elif __mode__ in ["cluster"]:
    base_path = Path(__file__).resolve().parent.parent

data_path = Path("/data1/sagawa/DirectMultiStep/data")

run_name = "reproduce_flex_duo_result_seed=42"
logbook_path = data_path / "configs" / "logbook" / run_name
train_conf = TrainingConfig.load(Path("/data1/sagawa/DirectMultiStep/data/configs/logbook/reproduce_flex_duo_result_seed=42/training_config.yaml"))
factory = ModelFactory.from_config_file(Path("/data1/sagawa/DirectMultiStep/data/configs/logbook/reproduce_flex_duo_result_seed=42/model_config.yaml"), compile_model=False)

ec = EvalConfig(
    data_path=data_path,
    run_name=run_name,
    eval_dataset="n1",
    epoch=40,
    use_sm=True,
    use_steps=True,
    beam_width=50,
    enc_active_experts=2,
    dec_active_experts=2,
    _checkpoint_path=Path("/data1/sagawa/DirectMultiStep/data/training/reproduce_flex_duo_result_seed=42/last.ckpt"),
)
ec.save(logbook_path / f"{ec.eval_name}_config.yaml")


if __name__ == "__main__":
    factory.check_for_eval_config_updates(ec)
    model = factory.create_model()
    device = ModelFactory.determine_device()
    model = factory.load_lightning_checkpoint(model, ec.checkpoint_path, device=device)

    evalObj = ModelEvaluator(model, ec, train_conf, device=device)

    evalObj.load_eval_dataset()
    evalObj.prepare_beam_search()

    all_beam_results_NS2 = evalObj.run_beam_search()

    top_ks = evalObj.calculate_top_k_accuracy()
    print(top_ks)
```

Evaluation on the n1 dataset:
```
solved (all): 10000
solved (valid): 9993
accuracy (valid, no perms):
  Top 1: '14.1'
  Top 2: '22.3'
  Top 3: '26.7'
  Top 4: '29.5'
  Top 5: '31.6'
  Top 10: '37.3'
  Top 20: '42.8'
  Top 50: '48.0'
accuracy (valid, with perms):
  Top 1: '24.0'
  Top 2: '28.4'
  Top 3: '31.8'
  Top 4: '34.4'
  Top 5: '36.5'
  Top 10: '42.0'
  Top 20: '47.2'
  Top 50: '52.5'
```

**Comment by anmorgunov** [comment]
hi, thanks for the interest, sorry for the delayed response.

the numbers look roughly correct, we used epoch 20 for flex 20M, here are the accuracy numbers based on valid paths

```yaml
solved (valid): 9995
accuracy (valid, no perms):
  Top 1: '18.3'
  Top 2: '28.4'
  Top 3: '34.4'
  Top 4: '37.6'
  Top 5: '40.1'
  Top 10: '46.7'
  Top 20: '51.9'
  Top 50: '57.5'
accuracy (valid, with perms):
  Top 1: '32.2'
  Top 2: '36.3'
  Top 3: '40.0'
  Top 4: '43.2'
  Top 5: '45.4'
  Top 10: '51.4'
  Top 20: '56.6'
  Top 50: '61.8'
```

please note that in the table we use top-k calculated after enforcing stock (in this case, following paroutes procedure, the set of all leaves from n1 or n5):

```yaml
solved (available): 8828
solved (correct): 8828
accuracy (processed, no perms):
  Top 1: '23.6'
  Top 2: '27.5'
  Top 3: '29.1'
  Top 4: '30.0'
  Top 5: '30.6'
  Top 10: '31.7'
  Top 20: '32.1'
  Top 50: '32.1'
accuracy (processed, with perms):
  Top 1: '43.2'
  Top 2: '51.7'
  Top 3: '55.3'
  Top 4: '57.2'
  Top 5: '58.4'
  Top 10: '60.8'
  Top 20: '61.7'
  Top 50: '61.8'
```

this comes after calling calculate_top_k_accuracy with `check_true_reacs=False` and  `check_stock=True`

we do mention that we enforce stock in sec 2.3.

> after beam search finds end-of-sequence tokens for all 50 candidates, predicted routes
are checked for validity of all SMILES strings (which are canonicalized), the presence of starting
materials in the stock set, and absence of repetitions (which can be present as permutations of other
routes)

lmk if you have any questions

**Comment by anmorgunov** [comment]
also, we used lr 3.5e-4 for Flex (see Sec 5.3.)

```py
    learning_rate=2e-4,
```

**Comment by sagawatatsuya** [comment]
Thank you for the reply!
I'll check the settings and try again.
